The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9647757764/job/26607341819) [(plain)](https://github.com/rust-lang/rust/commit/53ca7a97cb4f38c565549d0cd123c978aeb5e018/checks/26607341819/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.432   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.450 Collecting boolean-py==4.0
#13 0.459   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.476 Collecting chardet==5.1.0
---
#13 3.729 Building wheels for collected packages: reuse
#13 3.730   Building wheel for reuse (pyproject.toml): started
#13 4.050   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.051   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.051   Stored in directory: /tmp/pip-ephem-wheel-cache-9m_19ptt/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.053 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.075   Attempting uninstall: setuptools
#13 4.075     Found existing installation: setuptools 59.6.0
#13 4.077     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.077     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.077     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.739 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.740 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.277 Collecting virtualenv
#13 5.324   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.523      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 29.0 MB/s eta 0:00:00
#13 5.578 Collecting platformdirs<5,>=3.9.1
#13 5.584   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.621 Collecting filelock<4,>=3.12.2
#13 5.627   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.647 Collecting distlib<1,>=0.3.7
#13 5.653   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.662      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 64.8 MB/s eta 0:00:00
#13 5.754 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.924 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      229312 kB
DirectMap2M:     8159232 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/06c072f158c91296a2f802f2a10c9fe2499f314d/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-06c072f158c91296a2f802f2a10c9fe2499f314d-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 26.82s
##[endgroup]
fmt check
##[error]Diff in /checkout/src/bootstrap/src/core/build_steps/llvm.rs at line 407:
             cfg.define("LLVM_LINK_LLVM_DYLIB", "ON");
 
 
-        if (target.starts_with("csky") || target.starts_with("riscv") || target.starts_with("sparc-"))
+        if (target.starts_with("csky")
+            || target.starts_with("riscv")
+            || target.starts_with("sparc-"))
             && !target.contains("freebsd")
             && !target.contains("openbsd")
             && !target.contains("netbsd")
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/bootstrap/src/utils/tarball.rs" "/checkout/src/bootstrap/src/utils/job.rs" "/checkout/src/bootstrap/src/utils/cc_detect.rs" "/checkout/src/bootstrap/src/utils/channel.rs" "/checkout/src/bootstrap/src/utils/cache.rs" "/checkout/src/bootstrap/src/utils/helpers/tests.rs" "/checkout/src/bootstrap/src/utils/dylib.rs" "/checkout/src/bootstrap/src/utils/bin_helpers.rs" "/checkout/src/bootstrap/src/utils/helpers.rs" "/checkout/src/bootstrap/src/lib.rs" "/checkout/src/bootstrap/src/bin/rustc.rs" "/checkout/src/bootstrap/src/bin/rustdoc.rs" "/checkout/src/bootstrap/src/bin/sccache-plus-cl.rs" "/checkout/src/bootstrap/src/bin/main.rs" "/checkout/src/bootstrap/src/core/sanity.rs" "/checkout/src/bootstrap/src/core/download.rs" "/checkout/src/bootstrap/src/core/config/tests.rs" "/checkout/src/bootstrap/src/core/config/flags.rs" "/checkout/src/bootstrap/src/core/config/mod.rs" "/checkout/src/bootstrap/src/core/config/config.rs" "/checkout/src/bootstrap/src/core/mod.rs" "/checkout/src/bootstrap/src/core/builder.rs" "/checkout/src/bootstrap/src/core/builder/tests.rs" "/checkout/src/bootstrap/src/core/metadata.rs" "/checkout/src/bootstrap/src/core/build_steps/clippy.rs" "/checkout/src/bootstrap/src/core/build_steps/llvm.rs" "/checkout/src/bootstrap/src/core/build_steps/clean.rs" "/checkout/src/bootstrap/src/core/build_steps/test.rs" "/checkout/src/bootstrap/src/core/build_steps/mod.rs" "/checkout/src/bootstrap/src/core/build_steps/check.rs" "/checkout/src/bootstrap/src/core/build_steps/tool.rs" "/checkout/src/bootstrap/src/core/build_steps/run.rs" "/checkout/src/bootstrap/src/core/build_steps/synthetic_targets.rs" "/checkout/src/bootstrap/src/core/build_steps/perf.rs" "/checkout/src/bootstrap/src/core/build_steps/format.rs" "/checkout/src/bootstrap/src/core/build_steps/vendor.rs" "/checkout/src/bootstrap/src/core/build_steps/compile.rs" "/checkout/src/bootstrap/src/core/build_steps/install.rs" "/checkout/src/bootstrap/src/core/build_steps/suggest.rs" "/checkout/src/bootstrap/src/core/build_steps/setup.rs" "/checkout/src/bootstrap/src/core/build_steps/doc.rs" "/checkout/src/bootstrap/src/core/build_steps/setup/tests.rs" "/checkout/src/bootstrap/src/core/build_steps/toolstate.rs" "/checkout/src/bootstrap/src/core/build_steps/dist.rs" "/checkout/library/sysroot/src/lib.rs" "/checkout/library/test/src/tests.rs" "/checkout/library/test/src/event.rs" "/checkout/library/test/src/test_result.rs" "/checkout/library/test/src/cli.rs" "/checkout/library/test/src/console.rs" "/checkout/library/test/src/term.rs" "/checkout/library/test/src/formatters/junit.rs" "/checkout/library/test/src/formatters/mod.rs" "/checkout/library/test/src/formatters/terse.rs" "/checkout/library/test/src/formatters/json.rs" "/checkout/library/test/src/formatters/pretty.rs" "/checkout/library/test/src/term/win.rs" "/checkout/library/test/src/term/terminfo/mod.rs" "/checkout/library/test/src/term/terminfo/parm.rs" "/checkout/library/test/src/term/terminfo/parm/tests.rs" "/checkout/library/test/src/term/terminfo/searcher/tests.rs" "/checkout/library/test/src/term/terminfo/searcher.rs" "/checkout/library/test/src/term/terminfo/parser/compiled/tests.rs" "/checkout/src/bootstrap/src/utils/render_tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Jun 24 15:10:36 UTC 2024
  network time: Mon, 24 Jun 2024 15:10:36 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_target/src/spec/targets/sparc_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/sparc_unknown_linux_gnu.rs
@@ -1,13 +1,7 @@
 use crate::abi::Endian;
-use crate::spec::{base, Cc, LinkerFlavor, Lld, Target};
+use crate::spec::{base, Cc, LinkerFlavor, Lld, Target, TargetOptions};
 
 pub fn target() -> Target {
-    let mut base = base::linux_gnu::opts();
-    base.endian = Endian::Big;
-    base.cpu = "v9".into();
-    base.max_atomic_width = Some(32);
-    base.add_pre_link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-mv8plus"]);
-
     Target {
         llvm_target: "sparc-unknown-linux-gnu".into(),
         metadata: crate::spec::TargetMetadata {
@@ -19,6 +13,15 @@ pub fn target() -> Target {
         pointer_width: 32,
         data_layout: "E-m:e-p:32:32-i64:64-f128:64-n32-S64".into(),
         arch: "sparc".into(),
-        options: base,
+        options: TargetOptions {
+            cpu: "v9".into(),
+            endian: Endian::Big,
+            late_link_args: TargetOptions::link_args(
+                LinkerFlavor::Gnu(Cc::Yes, Lld::No),
+                &["-mv8plus"],
+            ),
+            max_atomic_width: Some(32),
+            ..base::linux_gnu::opts()
+        },
     }
 }

--- a/compiler/rustc_target/src/spec/targets/sparc_unknown_linux_gnu.rs
+++ b/compiler/rustc_target/src/spec/targets/sparc_unknown_linux_gnu.rs
@@ -18,7 +18,7 @@ pub fn target() -> Target {
             endian: Endian::Big,
             late_link_args: TargetOptions::link_args(
                 LinkerFlavor::Gnu(Cc::Yes, Lld::No),
-                &["-mv8plus"],
+                &["-mcpu=v9", "-m32"],
             ),
             max_atomic_width: Some(32),
             ..base::linux_gnu::opts()

--- a/src/bootstrap/src/core/build_steps/llvm.rs
+++ b/src/bootstrap/src/core/build_steps/llvm.rs
@@ -407,18 +407,21 @@ impl Step for Llvm {
             cfg.define("LLVM_LINK_LLVM_DYLIB", "ON");
         }
 
-        if (target.starts_with("riscv") || target.starts_with("csky"))
+        if (target.starts_with("csky")
+            || target.starts_with("riscv")
+            || target.starts_with("sparc-"))
             && !target.contains("freebsd")
             && !target.contains("openbsd")
             && !target.contains("netbsd")
         {
-            // RISC-V and CSKY GCC erroneously requires linking against
+            // CSKY and RISC-V GCC erroneously requires linking against
             // `libatomic` when using 1-byte and 2-byte C++
             // atomics but the LLVM build system check cannot
             // detect this. Therefore it is set manually here.
             // Some BSD uses Clang as its system compiler and
             // provides no libatomic in its base system so does
-            // not want this.
+            // not want this. 32-bit SPARC requires linking against
+            // libatomic as well.
             ldflags.exe.push(" -latomic");
             ldflags.shared.push(" -latomic");
         }

